src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -388,15 +461,35 @@ private bool ShouldInvalidateServer(Exception exception)
                         code = (ServerErrorCode)writeConcernError.GetValue("code", -1).ToInt32();
                         message = writeConcernError.GetValue("errmsg", null)?.AsString;
 
-                        if (IsNotMasterOrRecovering(code, message))
+                        if (IsStateChangeError(code, message))
```

Comment:
Loosely equivalent to these lines in the [spec commit](https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1365-R1378)